taqueria-sdk/index.ts

```diff
@@ -73,6 +75,21 @@ export const execCmd = (cmd: string): LikeAPromise<StdIO, ExecException> =>
 		});
 	});
 
+export const spawnCmd = (fullCmd: CmdArgEnv): Promise<StdIO> =>
+	new Promise((resolve, reject) => {
```

Comment:
@jchenche nice to see a spawnCmd method added to the SDK.

However, there are a few problems. 

I think we should first define an intent for this method: "spawnCmd is used to perform non-blocking with minimal (if any) buffering, and do not require independent pipes for stdout/stderr/stdin".

`child_process.spawn` can also be used for launching child processes that have custom processing needs, but I think we should focus our method for the use-case when "independent pipes" aren't required, as developers can always import `child_process.spawn` directly.

After refining the intent of `spawnCmd`, there are some problems we'll find with our implementation:

## spawnSync is used instead of spawn.
This has a few problems in itself:
1) the event loop is blocked. This means that other microtasks, such as any promises waiting to be resolved, or any internal asynchronous microtasks started by the runtime, must wait till spawnSync's child process is closed. Looking at the definition is our function, "waiting" is not permitted here. Concretely, if you look at our use-case, there's no desire for any kind of blocking I/O.

2) Mixing asynchronously and synchronous code is frowned upon as you have to be so conscious not to create incident logic errors.  See this code:
```
const fs = require('fs');
fs.readFile('/file.md', (err, data) => {
  if (err) throw err;
  console.log(data);
});
fs.unlinkSync('/file.md');
```

This kind of thing is hard to catch in a code review, as you have to keep track in your head, "is this using section of the code using async operations or sync operations?". That kind of cognitive load is highly undesirable and unproductive, and thus, its a best practice to use just one or the other, and in the majority of use-cases, async is usually want is wanted.

## We're buffering ALL output from the stdout and stderr pipes:
`child_process.spawn` will by default, create pipes for stdout/stderr/stdin, and have internal buffers used to capture the data microtask: `on.('data', buffer => handler(buffer))`. The data microtask can produce new buffers of data each time IPC data is sent from the child process, meaning there are independent buffers each time there is new data to process, and these buffers also have independent lifecycles so the garbage collector can ran rather optimal.

However, we're calling .toString() on both of the open pipes:
```
resolve({
    stdout: child.stdout ? child.stdout.toString() : '',
    stderr: child.stderr ? child.stderr.toString() : '',
});
```

By calling the .toString() method on a pipe, we're effectively doing the following:
```
return new Promise((resolve, reject) => {
  const stdout = fnWhichSpawnsAndReturnsStdoutPipe();
  const everythingBuffered  = []
  stdout.on('data', data => everythingBuffered.push(data.toString()))
  stdout.on('close', exitCode => exitCode === 0 ? resolve(everythingBuffered.join('')) : reject(exitCode))
});
```

# Recommendation:
- Use `spawn` instead of `spawnSync`.
- `spawnCmd` should resolve with the statusCode, or reject on error:
```
child.on('exit', resolve)
child.on('exit', reject)
```
- I believe we can use the `stdio` configuration option to control what pipes we open. After quickly looking at the docs, I think you can use `'inherit'` as the value which will ensure that the parent and child processes just share the same stdout/stderr/stdin pipes. If that doesn't work, then we just should just pipe the child's stream to the parent.

Reply:
Thanks Michael will use spawn.